internal/journal.go

```diff
@@ -0,0 +1,227 @@
+	if err != nil {
+		return err
+	}
+
+	allBackupsInfo[sentinelName] = info
+
+	err = UpdateBackupsInfo(folder, allBackupsInfo)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func UpdatePreviousBackupInfoJournal(folder storage.Folder, journalPath string, newJournalEnd string) error {
+	lastSentinelName, lastSentinel, err := GetLastNotPermanentBackupInfo(folder)
+	if err != nil {
+		return err
+	}
+
+	if len(lastSentinelName) == 0 {
+		tracelog.WarningLogger.Printf("last sentinel was not found, we can not evaluate journal size")
+		return nil
+	}
+
+	journalSize, err := GetJournalSizeInSemiInterval(
+		folder,
+		journalPath,
+		func(a, b string) bool {
+			return a < b
+		},
+		lastSentinel.JournalEnd,
+		newJournalEnd,
+	)
+	if err != nil {
+		tracelog.ErrorLogger.Printf("can not evaluate journal sum for %s: %s", lastSentinelName, err)
+		return err
+	}
+	tracelog.InfoLogger.Printf(
+		"journal size for %s in the semi interval (%s; %s] is equal to %d",
+		lastSentinelName,
+		lastSentinel.JournalEnd,
+		newJournalEnd,
+		journalSize,
+	)
+
+	lastBackupInfo, err := GetBackupInfo(folder, lastSentinelName)
+	if err != nil {
+		tracelog.ErrorLogger.Printf("can not find journal sum in backups.json for %s: %s", lastSentinelName, err)
+		return err
+	}
+
+	if lastBackupInfo.JournalSize != 0 {
+		tracelog.WarningLogger.Printf(
+			"previous backup info contains non-zero journal size '%d', its values will be updated to '%d'",
+			lastBackupInfo.JournalSize,
+			journalSize,
+		)
+	}
+
+	err = UploadBackupInfo(folder, lastSentinelName, BackupInfo{
+		JournalStart:     lastSentinel.JournalStart,
+		JournalEnd:       lastSentinel.JournalEnd,
+		JournalSize:      journalSize,
+		CompressedSize:   lastSentinel.CompressedSize,
+		UncompressedSize: lastSentinel.UncompressedSize,
+		IsPermanent:      lastSentinel.IsPermanent,
+		StopLocalTime:    lastSentinel.StopLocalTime,
+	})
+	if err != nil {
+		tracelog.ErrorLogger.Printf("can not update journal info for %s: %s", lastSentinelName, err)
+		return err
+	}
+
+	tracelog.ErrorLogger.Printf("journal info has been updated for %s", lastSentinelName)
+	return nil
+}
+
+func GetLastNotPermanentBackupInfo(folder storage.Folder) (string, BackupInfo, error) {
+	allBackupsInfo, err := GetAllBackupsInfo(folder)
+	if err != nil {
+		return "", BackupInfo{}, err
+	}
+
+	lastBackupInfo := BackupInfo{}
+	lastBackupName := ""
+
+	for k, v := range allBackupsInfo {
+		if !v.IsPermanent && v.StopLocalTime.After(lastBackupInfo.StopLocalTime) {
+			lastBackupName = k
+			lastBackupInfo = v
+		}
+	}
+
+	return lastBackupName, lastBackupInfo, nil
+}
+
+func GetAllBackupsInfo(folder storage.Folder) (map[string]BackupInfo, error) {
+	ok, err := folder.Exists(BackupsInfo)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		tracelog.InfoLogger.Printf("can not find backups.json, creating it...")
+		err := folder.PutObject(BackupsInfo, bytes.NewBuffer([]byte("{}")))
+		if err != nil {
+			return nil, err
+		}
```

Comment:
It seems that you can skip empty json creation - just return empty map to caller.

Reply:
I think we could add logging to the error section and retries for the S3 functions like PutObject, ReadObject, and ListFolder.